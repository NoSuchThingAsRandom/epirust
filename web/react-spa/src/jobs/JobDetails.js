/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import React from "react";

import "./job-details.scss"
import { NavItem } from "../common/NavItem";
import { Redirect, Route, Switch } from "react-router-dom";
import GridPage from "../grid";
import PropTypes from 'prop-types';
import { TimeSeries } from "../time-series";
import { TimeSeriesDeviation } from "../time-series/TimeSeriesDeviation";

export const JobDetails = ({ jobId, details }) => {
  const linkPrefix = `/jobs/${jobId}`;

  const isGridEnabled = details && details.config && details.config.enable_citizen_state_messages || false,
    isFinished = details && details.status === 'finished';

  function renderGraphTabs() {
    return (
      <div className="row">
        <div className="col-8">
          <ul className="nav nav-tabs nav-fill">

            <NavItem name="Time Series" linksTo={`${linkPrefix}/time-series`} activeOnExactMatch={true} />
            {isFinished && <NavItem name="Time Series Deviation" linksTo={`${linkPrefix}/time-series-deviation`} />}
            {isGridEnabled && <NavItem name="Grid" linksTo={`${linkPrefix}/grid`} />}
            <NavItem name="Config" linksTo={`${linkPrefix}/config`} />

          </ul>
        </div>

        <div className="col-4">
          <h4 className="simulation-id">{`Simulation: `}<code>{jobId}</code></h4>
        </div>
      </div>
    );
  }

  function renderContentForTab() {
    return (
      <div className='job-details-content'>
        <Switch>

          <Route exact path={"/jobs/:id/time-series"}>
            <TimeSeries jobId={jobId} />
          </Route>

          <Route exact path={"/jobs/:id/time-series-deviation"}>
            <TimeSeriesDeviation jobId={jobId} />
          </Route>

          <Route exact path={"/jobs/:id/grid"}>
            {isGridEnabled ? <GridPage /> : <Redirect to={`/jobs/${jobId}/time-series`} />}
          </Route>

          <Route exact path={"/jobs/:id/config"}>
            {details && <pre>{JSON.stringify(details.config, undefined, 4)}</pre>}
          </Route>
        </Switch>
      </div>
    );
  }


  return (
    <div className="job-details">
      {renderGraphTabs()}
      {renderContentForTab()}
    </div>
  )
};

JobDetails.propTypes = {
  jobId: PropTypes.string.isRequired,
  details: PropTypes.object
};
