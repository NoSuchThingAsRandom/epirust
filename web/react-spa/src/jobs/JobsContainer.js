/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import {Redirect} from "react-router-dom";
import {Jobs} from "./Jobs";
import React from "react";

export const JobsContainer = ({activeJobId, currentView, jobs}) => {
  if (!activeJobId)
    return (<Redirect to={`/jobs/${jobs[0]._id}/time-series`}/>);
  if (!currentView)
    return (<Redirect to={`/jobs/${activeJobId}/time-series`}/>);

  const activeJob = jobs.find(j => j._id === activeJobId)
  if (!activeJob) return <div>Invalid job requested</div>
  else return <Jobs jobs={jobs} activeJob={activeJob}/>
}
