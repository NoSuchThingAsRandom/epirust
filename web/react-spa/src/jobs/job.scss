/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

.borderless {
  border: none;
}

.card-body-less-padding {
  padding: 0.5rem;
}

.item-less-padding {
  padding: 0.25rem 1.25rem
}

.job-status {
  font-size: 22px;
  font-weight: bold;
  line-height: 26px;
}

.minor-details {
  font-size: 16px;
}

.no-link-formatting {
  color: white;
  text-decoration: none;
  &:focus, &:hover, &:visited, &:link, &:active {
    text-decoration: none;
    color: white;
  }
}