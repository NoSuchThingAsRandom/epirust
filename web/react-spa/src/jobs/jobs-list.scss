/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

.jobs-list {
  .tabs {
  }
  .details {
    padding: 0 20px;
  }
}

.scrollable {
  height: 90vh;
  position: relative;
  overflow: auto;
}

.running {
  background-color: #F6BB43;
}

.failed {
  background-color: #F9694C;
}

.finished {
  background-color: #8BC253;
}

.in-queue {
  background-color: #3CAEDA;
}

.left-border {
  border-style: none none none solid;
  border-color: black;
}

.padded-list {
  padding: 0 16px 0 0;
}