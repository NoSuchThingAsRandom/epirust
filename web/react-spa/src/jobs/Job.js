/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import React from "react";
import './job.scss'
import PropTypes from "prop-types";
import { Link } from "react-router-dom";

export const Job = ({ simulationId, status, isActive = false }) => {
  const statusText = {
    "running": "In-Progress",
    "finished": "Finished",
    "failed": "Failed",
    "in-queue": "In-Queue"
  };

  return (
    <li className="list-group-item borderless item-less-padding">

      <Link to={`/jobs/${simulationId}`} className={'no-link-formatting'}>
        <div className={`simulation-tab ${isActive ? "active shadow" : ""} ${status}`}>
          <div className="card-body card-body-less-padding">
            <p data-testid={`job-status-${simulationId}`} className="job-status">{statusText[status]}</p>
            <p className="minor-details">{`Id: `}<code>{simulationId}</code></p>
          </div>
        </div>
      </Link>

    </li>
  );
};

Job.propTypes = {
  simulationId: PropTypes.number.isRequired,
  status: PropTypes.oneOf(["running", "finished", "failed", "in-queue"]).isRequired
};