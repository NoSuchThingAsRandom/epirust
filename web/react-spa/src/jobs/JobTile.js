/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import React from "react";
import '../styles/components/jobs/job.scss'
import PropTypes from "prop-types";
import { Link } from "react-router-dom";

export const JobTile = ({jobId, status, isActive }) => {
  const statusText = {
    "running": "In-Progress",
    "finished": "Finished",
    "failed": "Failed",
    "in-queue": "In-Queue"
  };

  return (
    <li className="list-group-item borderless item-less-padding">

      <Link to={`/jobs/${jobId}`} className={'no-link-formatting'}>
        <div className={`simulation-tab ${isActive ? "active shadow" : ""} ${status}`}>
          <div className="card-body card-body-less-padding">
            <p data-testid={`job-status-${jobId}`} className="job-status">{statusText[status]}</p>
            <p className="minor-details">{`Id: `}<code>{jobId}</code></p>
          </div>
        </div>
      </Link>

    </li>
  );
};

JobTile.propTypes = {
  jobId: PropTypes.string.isRequired,
  status: PropTypes.oneOf(["running", "finished", "failed", "in-queue"]).isRequired
};