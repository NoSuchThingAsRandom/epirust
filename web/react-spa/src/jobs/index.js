/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import React, {useEffect, useState} from "react";
import '../styles/components/jobs/jobs-list.scss'
import {useParams} from 'react-router-dom';
import {get} from "../common/apiCall";
import {reduceStatus} from "./JobTransformer";
import {LOADING_STATES} from "../common/constants";
import {JobsContainer} from "./JobsContainer";
import {LoadingComponent} from "../common/LoadingComponent";

export const JobsView = () => {
  const [loadingState, updateLoadingState] = useState(LOADING_STATES.LOADING)
  const [jobs, updateJobs] = useState([])
  const {id: activeJobId, view} = useParams();

  const refreshJobs = (jobsToProcess) => {
    const jobsToFetch = jobsToProcess.filter(j => j.status !== "finished")
    if (jobsToFetch.length > 0) {
      const url = `/jobs?jobIds=${jobsToFetch.map(j => j._id).join(",")}`
      get(url)
        .then(res => res.json())
        .then(jobsToUpdate => {
          updateJobs(prevJobs => {
            const updatedJobs = prevJobs.map(pj => {
              const jobToUpdate = jobsToUpdate.find(ju => ju._id === pj._id);
              return jobToUpdate ? reduceStatus(jobToUpdate) : pj;
            });

            setTimeout(() => refreshJobs(updatedJobs), 15000);
            return updatedJobs;
          })
        })
    }
  }

  useEffect(() => {
    get("/jobs")
      .then(res => res.json())
      .then(receivedJobs => {
        const convertedJobs = receivedJobs.map(reduceStatus).reverse();
        updateJobs(convertedJobs)
        updateLoadingState(LOADING_STATES.FINISHED)
        setTimeout(() => refreshJobs(convertedJobs), 15000)
      })
      .catch(() => updateLoadingState(LOADING_STATES.FAILED))
  }, []);

  const children = <JobsContainer activeJobId={activeJobId} currentView={view} jobs={jobs}/>;
  return <LoadingComponent loadingState={loadingState} children={children}/>;
}
