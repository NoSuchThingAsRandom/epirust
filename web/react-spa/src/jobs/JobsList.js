/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import React, {useEffect, useState} from "react";
import './jobs-list.scss'
import {Job} from "./Job";
import {JobDetails} from "./JobDetails";
import {Redirect, useParams} from 'react-router-dom';

export const JobsList = () => {
  const {id, view} = useParams();
  const [simulations, updateSimulations] = useState([]);

  useEffect(() => {
    fetch("http://localhost:3000/simulation/",
      {
        method: 'GET',
        headers: {'Content-Type': 'application/json'}
      })
      .then(value => value.json())
      .then(value => updateSimulations(value.reverse()))
  }, []);

  if (id && !view) {
    return (<Redirect to={`/jobs/${id}/time-series`}/>);
  } else {
    return (<div className="row jobs-list">
      <div className="col-3">
        <ul className="list-group scrollable">
          {simulations.map(s => <Job key={s.simulation_id} simulationId={s.simulation_id} status={s.status}/>)}
        </ul>
      </div>
      <div className="col-9">
        {id && <JobDetails simulationId={id}/>}
      </div>
    </div>);
  }
};
