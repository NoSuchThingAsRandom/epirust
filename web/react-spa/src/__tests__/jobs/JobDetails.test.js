/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import {JobDetails} from "../../jobs/JobDetails";
import {render} from "@testing-library/react";
import React from "react";
import {MemoryRouter} from "react-router-dom";

describe('Job Details', () => {
    it('should render the time series if the view is time-series', () => {
        const component = render(<MemoryRouter initialEntries={['/jobs/123/time-series']}>
            <JobDetails jobId={'123'} details={{ config: {} }} />
        </MemoryRouter>);
        //TODO: Find a better way to assert. Changes irrelevant to the test would fail this test
        expect(component.container).toMatchSnapshot()
    });

    it('should render the config if the config is given', () => {
        const component = render(<MemoryRouter initialEntries={['/jobs/123/config']}>
            <JobDetails jobId={'123'} details={{ config: { configKey: 'configValue' } }} />
        </MemoryRouter>);

        expect(component.container).toMatchSnapshot()
    });

    it('should not render the config if the config is not given', () => {
        const component = render(<MemoryRouter initialEntries={['/jobs/123/config']}>
            <JobDetails jobId={'123'} />
        </MemoryRouter>);

        expect(component.container).toMatchSnapshot()
    });

    it('should render loader initially when no info received from socket', () => {
        const simulationDetails = { config: { enable_citizen_state_messages: false } };
        const { asFragment } = render(<MemoryRouter initialEntries={['/jobs/123/grid']}>
            <JobDetails jobId={'123'} details={simulationDetails} />
        </MemoryRouter>);

        expect(asFragment()).toMatchSnapshot()
    });

    it('should render grid if grid is enabled in config', () => {
        const simulationDetails = { config: { enable_citizen_state_messages: true } };
        const component = render(<MemoryRouter initialEntries={['/jobs/123/grid']}>
            <JobDetails jobId={'123'} details={simulationDetails} />
        </MemoryRouter>);

        expect(component.container).toMatchSnapshot()
    });
});