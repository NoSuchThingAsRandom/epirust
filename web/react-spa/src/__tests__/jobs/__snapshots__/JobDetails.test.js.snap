// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Job Details should not render the config if the config is not given 1`] = `
<div>
  <div
    class="job-details"
    style="padding-top: 4px;"
  >
    <h4>
      Simulation: 
      123
    </h4>
    <ul
      class="nav nav-tabs"
    >
      <li
        class="nav-item"
      >
        <a
          class="nav-link "
          href="/jobs/123/time-series"
        >
          Time Series
        </a>
      </li>
      <li
        class="nav-item"
      >
        <a
          class="nav-link active"
          href="/jobs/123/config"
        >
          Config
        </a>
      </li>
    </ul>
  </div>
</div>
`;

exports[`Job Details should render grid if grid is enabled in config 1`] = `
<div>
  <div
    class="job-details"
    style="padding-top: 4px;"
  >
    <h4>
      Simulation: 
      123
    </h4>
    <ul
      class="nav nav-tabs"
    >
      <li
        class="nav-item"
      >
        <a
          class="nav-link "
          href="/jobs/123/time-series"
        >
          Time Series
        </a>
      </li>
      <li
        class="nav-item"
      >
        <a
          class="nav-link active"
          href="/jobs/123/grid"
        >
          Grid
        </a>
      </li>
      <li
        class="nav-item"
      >
        <a
          class="nav-link "
          href="/jobs/123/config"
        >
          Config
        </a>
      </li>
    </ul>
    <div
      style="position: relative;"
    >
      <canvas
        data-testid="grid-canvas"
        height="701"
        id="grid-canvas"
        style="position: absolute; z-index: 1;"
        width="701"
      />
      <canvas
        data-testid="grid-canvas-landmarks"
        height="701"
        id="grid-canvas-landmarks"
        style="position: absolute; z-index: 3;"
        width="701"
      />
      <div
        style="position: relative;"
      >
        <div
          style="position: absolute; z-index: 5; right: 0px;"
        >
          <h4>
            0/0 hrs
          </h4>
          <button
            class="btn btn-success"
          >
            START
          </button>
          <button
            class="btn btn-danger"
          >
            STOP
          </button>
        </div>
        <canvas
          height="701"
          id="grid-canvas-agents"
          style="position: absolute; z-index: 4;"
          width="701"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Job Details should render the config if the config is given 1`] = `
<div>
  <div
    class="job-details"
    style="padding-top: 4px;"
  >
    <h4>
      Simulation: 
      123
    </h4>
    <ul
      class="nav nav-tabs"
    >
      <li
        class="nav-item"
      >
        <a
          class="nav-link "
          href="/jobs/123/time-series"
        >
          Time Series
        </a>
      </li>
      <li
        class="nav-item"
      >
        <a
          class="nav-link active"
          href="/jobs/123/config"
        >
          Config
        </a>
      </li>
    </ul>
    <pre>
      {
    "configKey": "configValue"
}
    </pre>
  </div>
</div>
`;

exports[`Job Details should render the time series if the view is time-series 1`] = `
<div>
  <div
    class="job-details"
    style="padding-top: 4px;"
  >
    <h4>
      Simulation: 
      123
    </h4>
    <ul
      class="nav nav-tabs"
    >
      <li
        class="nav-item"
      >
        <a
          class="nav-link active"
          href="/jobs/123/time-series"
        >
          Time Series
        </a>
      </li>
      <li
        class="nav-item"
      >
        <a
          class="nav-link "
          href="/jobs/123/config"
        >
          Config
        </a>
      </li>
    </ul>
    <div
      class="graph-wrap"
    >
      <div
        data-testid="visualization"
        id="vis"
      />
      <div
        class="graph-actions"
      >
        <p
          class="action"
        >
          Zoom: 
          <span
            class="control"
          >
            click-drag
          </span>
        </p>
        <p
          class="action"
        >
          Pan: 
          <span
            class="control"
          >
            shift-click-drag
          </span>
        </p>
        <p
          class="action"
        >
          Restore: 
          <span
            class="control"
          >
            double-click
          </span>
        </p>
        <button
          class="btn btn-secondary"
          disabled=""
        >
          Export graph
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Job Details should render time-series if grid is disabled in config 1`] = `
<div>
  <div
    class="job-details"
    style="padding-top: 4px;"
  >
    <h4>
      Simulation: 
      123
    </h4>
    <ul
      class="nav nav-tabs"
    >
      <li
        class="nav-item"
      >
        <a
          class="nav-link active"
          href="/jobs/123/time-series"
        >
          Time Series
        </a>
      </li>
      <li
        class="nav-item"
      >
        <a
          class="nav-link "
          href="/jobs/123/config"
        >
          Config
        </a>
      </li>
    </ul>
    <div
      class="graph-wrap"
    >
      <div
        data-testid="visualization"
        id="vis"
      />
      <div
        class="graph-actions"
      >
        <p
          class="action"
        >
          Zoom: 
          <span
            class="control"
          >
            click-drag
          </span>
        </p>
        <p
          class="action"
        >
          Pan: 
          <span
            class="control"
          >
            shift-click-drag
          </span>
        </p>
        <p
          class="action"
        >
          Restore: 
          <span
            class="control"
          >
            double-click
          </span>
        </p>
        <button
          class="btn btn-secondary"
          disabled=""
        >
          Export graph
        </button>
      </div>
    </div>
  </div>
</div>
`;
